core/server/worker/src/main/java/alluxio/worker/netty/UfsFileWriteHandler.java

```diff
@@ -0,0 +1,227 @@
+
+  /**
+   * The UFS File write request internal representation.
+   */
+  @NotThreadSafe
+  static final class UfsFileWriteRequest extends WriteRequest {
+    private final String mUfsPath;
+    private final Protocol.CreateUfsFileOptions mCreateUfsFileOptions;
+    private UnderFileSystem mUnderFileSystem;
+    private OutputStream mOutputStream;
+    private Counter mCounter;
+
+    UfsFileWriteRequest(Protocol.WriteRequest request) {
+      super(request.getId());
+      mUfsPath = request.getCreateUfsFileOptions().getUfsPath();
+      mCreateUfsFileOptions = request.getCreateUfsFileOptions();
+    }
+
+    /**
+     * @return the UFS path
+     */
+    public String getUfsPath() {
+      return mUfsPath;
+    }
+
+    /**
+     * @return the options to create UFS file
+     */
+    public Protocol.CreateUfsFileOptions getCreateUfsFileOptions() {
+      return mCreateUfsFileOptions;
+    }
+
+    /**
+     * @return the output stream
+     */
+    @Nullable
+    public OutputStream getOutputStream() {
+      return mOutputStream;
+    }
+
+    /**
+     * @return the handler of the UFS
+     */
+    @Nullable
+    public UnderFileSystem getUnderFileSystem() {
+      return mUnderFileSystem;
+    }
+
+    /**
+     * @return the counter
+     */
+    @Nullable
+    public Counter getCounter() {
+      return mCounter;
+    }
+
+    /**
+     * @param outputStream output stream to set
+     */
+    public void setOutputStream(OutputStream outputStream) {
+      mOutputStream = outputStream;
+    }
+
+    /**
+     * @param underFileSystem UFS to set
+     */
+    public void setUnderFileSystem(UnderFileSystem underFileSystem) {
+      mUnderFileSystem = underFileSystem;
+    }
+
+    /**
+     * @param counter counter to set
+     */
+    public void setCounter(Counter counter) {
+      mCounter = counter;
+    }
+  }
+
+  /**
+   * Creates an instance of {@link UfsFileWriteHandler}.
+   *
+   * @param executorService the executor service to run {@link PacketWriter}s
+   * @param ufsManager the file data manager
+   */
+  UfsFileWriteHandler(ExecutorService executorService, UfsManager ufsManager) {
+    super(executorService);
+    mUfsManager = ufsManager;
+  }
+
+  @Override
+  protected boolean acceptMessage(Object object) {
+    if (!super.acceptMessage(object)) {
+      return false;
+    }
+    Protocol.WriteRequest request = ((RPCProtoMessage) object).getMessage().asWriteRequest();
+    return request.getType() == Protocol.RequestType.UFS_FILE;
+  }
+
+  @Override
+  protected UfsFileWriteRequest createRequest(RPCProtoMessage msg) throws Exception {
+    return new UfsFileWriteRequest(msg.getMessage().asWriteRequest());
+  }
+
+  @Override
+  protected void completeRequest(Channel channel) throws Exception {
+    UfsFileWriteRequest request = getRequest();
+    if (request == null) {
+      return;
+    }
+    if (request.getOutputStream() == null) {
+      createUfsFile(channel);
+    }
+    if (request.getOutputStream() != null) {
```

Comment:
Is it expected that `request.getOutputStream()` can still be `null` after `createUfsFile()`?